File: crates/spfs/spfs/storage/_repository_test.py
from typing import Iterable, Any
import stat
import pytest
import py.path

from .. import tracking
from . import fs, tar
from ._repository import Repository
from ._layer import Layer
from ._manifest import Manifest, ManifestViewer


@pytest.fixture(params=["fs", "tar"])
def tmprepo(request: Any, tmpdir: py._path.local.LocalPath) -> Iterable[Repository]:

    if request.param == "fs":
        tmpdir = tmpdir.join("repo")
        yield fs.FSRepository(tmpdir.strpath, create=True)
    else:
        yield tar.TarRepository(tmpdir.join("repo.tar").strpath)


def test_find_aliases(tmpdir: py.path.local, tmprepo: Repository) -> None:

    tmpdir = tmpdir.join("repo")
    tmprepo = fs.FSRepository(tmpdir.strpath, create=True)
    with pytest.raises(ValueError):
        tmprepo.find_aliases("not-existant")

    tmpdir.join("data", "file.txt").ensure()
    manifest = tmprepo.commit_dir(tmpdir.join("data").strpath)
    layer = tmprepo.create_layer(Manifest(manifest))
    tmprepo.tags.push_tag("test-tag", layer.digest())

    assert tmprepo.find_aliases(layer.digest().str()) == ["test-tag"]
    assert tmprepo.find_aliases("test-tag") == [layer.digest().str()]


def test_commit_mode(tmpdir: py.path.local, tmprepo: Repository) -> None:

    datafile_path = "dir1.0/dir2.0/file.txt"
    symlink_path = "dir1.0/dir2.0/file2.txt"

    src_dir = tmpdir.join("source")
    link_dest = src_dir.join(datafile_path)
    link_dest.write("somedata", ensure=True)
    src_dir.join(symlink_path).mksymlinkto(link_dest)
    link_dest.chmod(0o444)

    manifest = tmprepo.commit_dir(src_dir.strpath)
    if not isinstance(tmprepo, ManifestViewer):
        pytest.skip("Nothing to test for repo that is not a ManifestViewer")
        return

    rendered_dir = tmprepo.render_manifest(Manifest(manifest))
    rendered_symlink = py.path.local(rendered_dir).join(symlink_path)
    assert stat.S_ISLNK(rendered_symlink.lstat().mode)

    symlink_entry = manifest.get_path(symlink_path)
    payloads = tmprepo.payloads
    assert isinstance(payloads, fs.FSPayloadStorage)
    symlink_blob = py.path.local(payloads._build_digest_path(symlink_entry.object))
    assert not stat.S_ISLNK(symlink_blob.lstat().mode)


def test_commit_broken_link(tmpdir: py.path.local, tmprepo: Repository) -> None:

    src_dir = tmpdir.join("source").ensure(dir=True)
    src_dir.join("broken-link").mksymlinkto("nonexistant")

    manifest = tmprepo.commit_dir(src_dir.strpath)
    assert manifest.get_path("broken-link") is not None


def test_commit_dir(tmpdir: py.path.local, tmprepo: Repository) -> None:

    src_dir = tmpdir.join("source")
    src_dir.join("dir1.0/dir2.0/file.txt").write("somedata", ensure=True)
    src_dir.join("dir1.0/dir2.1/file.txt").write("someotherdata", ensure=True)
    src_dir.join("dir2.0/file.txt").write("evenmoredata", ensure=True)
    src_dir.join("file.txt").write("rootdata", ensure=True)

    manifest = Manifest(tmprepo.commit_dir(src_dir.strpath))
    manifest2 = Manifest(tmprepo.commit_dir(src_dir.strpath))
    assert manifest.digest() == manifest2.digest()

File: crates/spfs/src/storage/repository_test.rs
// Copyright (c) Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use std::collections::HashSet;
use std::iter::FromIterator;
use std::os::unix::fs::{MetadataExt, PermissionsExt};
use std::sync::Arc;

use rstest::rstest;

use super::Ref;
use crate::encoding::Encodable;
use crate::fixtures::*;
use crate::graph::Manifest;
use crate::storage::fs;
use crate::storage::prelude::*;
use crate::tracking::TagSpec;

#[rstest]
#[case::fs(tmprepo("fs"))]
#[case::tar(tmprepo("tar"))]
#[cfg_attr(feature = "server", case::rpc(tmprepo("rpc")))]
#[tokio::test]
async fn test_find_aliases(
    #[case]
    #[future]
    tmprepo: TempRepo,
) {
    init_logging();
    let tmprepo = tmprepo.await;
    tmprepo
        .find_aliases("not-existent")
        .await
        .expect_err("should error when ref is not found");

    let manifest = crate::Committer::new(&tmprepo)
        .commit_dir("src/storage")
        .await
        .unwrap();
    let layer = tmprepo
        .create_layer(&Manifest::from(&manifest))
        .await
        .unwrap();
    let test_tag = TagSpec::parse("test-tag").unwrap();
    tmprepo
        .push_tag(&test_tag, &layer.digest().unwrap())
        .await
        .unwrap();

    let actual = tmprepo
        .find_aliases(layer.digest().unwrap().to_string().as_ref())
        .await
        .unwrap();
    let expected = HashSet::from_iter(vec![Ref::TagSpec(test_tag)]);
    assert_eq!(actual, expected);
    let actual = tmprepo.find_aliases("test-tag").await.unwrap();
    let expected = HashSet::from_iter(vec![Ref::Digest(layer.digest().unwrap())]);
    assert_eq!(actual, expected);
}

#[rstest]
#[tokio::test]
async fn test_commit_mode_fs(tmpdir: tempfile::TempDir) {
    init_logging();
    let dir = tmpdir.path();
    let tmprepo = Arc::new(
        fs::FSRepository::create(dir.join("repo"))
            .await
            .unwrap()
            .into(),
    );
    let datafile_path = "dir1.0/dir2.0/file.txt";
    let symlink_path = "dir1.0/dir2.0/file2.txt";

    let src_dir = dir.join("source");
    std::fs::create_dir_all(src_dir.join("dir1.0/dir2.0")).unwrap();
    let link_dest = src_dir.join(datafile_path);
    std::fs::write(&link_dest, "somedata").unwrap();
    std::os::unix::fs::symlink(&link_dest, src_dir.join(symlink_path)).unwrap();
    std::fs::set_permissions(&link_dest, std::fs::Permissions::from_mode(0o444)).unwrap();

    let manifest = crate::Committer::new(&tmprepo)
        .commit_dir(&src_dir)
        .await
        .expect("failed to commit dir");

    // Safety: tmprepo was created as an FSRepository
    let tmprepo = match &*tmprepo {
        RepositoryHandle::FS(fs) => fs.opened().await.unwrap(),
        _ => panic!("Unexpected tmprepo type!"),
    };

    let rendered_dir = fs::Renderer::new(&*tmprepo)
        .render_manifest(&Manifest::from(&manifest), None)
        .await
        .expect("failed to render manifest");
    let rendered_symlink = rendered_dir.join(symlink_path);
    let rendered_mode = rendered_symlink.symlink_metadata().unwrap().mode();
    assert!(unix_mode::is_symlink(rendered_mode), "should be a symlink");

    let symlink_entry = manifest
        .get_path(symlink_path)
        .expect("symlink not in manifest");
    let symlink_blob = tmprepo.payloads.build_digest_path(&symlink_entry.object);
    let blob_mode = symlink_blob.symlink_metadata().unwrap().mode();
    assert!(
        !unix_mode::is_symlink(blob_mode),
        "stored blob should not be a symlink"
    )
}

#[rstest]
#[case::fs(tmprepo("fs"))]
#[case::tar(tmprepo("tar"))]
#[cfg_attr(feature = "server", case::rpc(tmprepo("rpc")))]
#[tokio::test]
async fn test_commit_broken_link(
    #[case]
    #[future]
    tmprepo: TempRepo,
    tmpdir: tempfile::TempDir,
) {
    let tmprepo = tmprepo.await;
    let src_dir = tmpdir.path().join("source");
    std::fs::create_dir_all(&src_dir).unwrap();
    std::os::unix::fs::symlink(
        std::path::Path::new("nonexistent"),
        src_dir.join("broken-link"),
    )
    .unwrap();

    let manifest = crate::Committer::new(&tmprepo)
        .commit_dir(&src_dir)
        .await
        .unwrap();
    assert!(manifest.get_path("broken-link").is_some());
}

#[rstest]
#[case::fs(tmprepo("fs"))]
#[case::tar(tmprepo("tar"))]
#[cfg_attr(feature = "server", case::rpc(tmprepo("rpc")))]
#[tokio::test]
async fn test_commit_dir(
    #[case]
    #[future]
    tmprepo: TempRepo,
    tmpdir: tempfile::TempDir,
) {
    let tmprepo = tmprepo.await;
    let src_dir = tmpdir.path().join("source");
    ensure(src_dir.join("dir1.0/dir2.0/file.txt"), "somedata");
    ensure(src_dir.join("dir1.0/dir2.1/file.txt"), "someotherdata");
    ensure(src_dir.join("dir2.0/file.txt"), "evenmoredata");
    ensure(src_dir.join("file.txt"), "rootdata");

    let manifest = Manifest::from(
        &crate::Committer::new(&tmprepo)
            .commit_dir(&src_dir)
            .await
            .unwrap(),
    );
    let manifest2 = Manifest::from(
        &crate::Committer::new(&tmprepo)
            .commit_dir(&src_dir)
            .await
            .unwrap(),
    );
    assert_eq!(manifest, manifest2);
}
